builtin/providers/aws/resource_aws_cloudwatch_logs_subscription_filter.go

```diff
@@ -0,0 +1,341 @@
+package aws
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/aws/aws-sdk-go/service/lambda"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+// Number of times to retry if a throttling- or test message exception occurs
+const CLOUDWATCH_LOGS_SUBSCRIPTION_FILTER_MAX_THROTTLE_RETRIES = 10
+
+// How long to sleep when a throttle-event happens
+const CLOUDWATCH_LOGS_SUBSCRIPTION_FILTER_THROTTLE_SLEEP_MILLISECONDS = 8000
+
+func resourceAwsCloudwatchLogsSubscriptionFilter() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCloudwatchLogsSubscriptionFilterCreate,
+		Read:   resourceAwsCloudwatchLogsSubscriptionFilterRead,
+		Update: resourceAwsCloudwatchLogsSubscriptionFilterUpdate,
+		Delete: resourceAwsCloudwatchLogsSubscriptionFilterDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"destination": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"filter_pattern": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"log_group": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"role": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCloudwatchLogsSubscriptionFilterCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudwatchlogsconn
+
+	name := d.Get("name").(string)
+
+	log_group := d.Get("log_group").(string)
+
+	destination := d.Get("destination").(string)
+	if strings.HasPrefix(destination, "arn:aws:kinesis:") {
+		destination_arn_sliced := strings.Split(destination, "/")
+		destination_name := destination_arn_sliced[len(destination_arn_sliced)-1]
+
+		kinesis_conn := meta.(*AWSClient).kinesisconn
+		waitForKinesisStreamToActivate(kinesis_conn, destination_name)
+	} else if strings.HasPrefix(destination, "arn:aws:lambda") {
+		lambda_conn := meta.(*AWSClient).lambdaconn
+
+		lambda_arn_sliced := strings.Split(destination, ":")
+		function_name := lambda_arn_sliced[len(lambda_arn_sliced)-1]
+		statement_id := lambdaPermissionStatementId(log_group, destination)
+
+		if !permissionExists(function_name, statement_id, lambda_conn) {
```

Comment:
I'm not sure it's a good idea to manage some resources "under the hood" and hide those away from the user. I think the user should be fully in control of all resources (and all parameters), where possible.

This is also bringing an unexpected behaviour - imagine a destructive update of the Lambda function (e.g. `runtime`) you rely on here - destroying the function would also destroy all lambda permissions and it would _not_ trigger recreation of the log subscription here (because the ARN would remain the same).

I will hopefully soon finish https://github.com/hashicorp/terraform/pull/4826 which should solve this problem in cleaner way - i.e. I think we should leave it up to the user to setup the Lambda Permission.